implementations/c/lib/memory/tests/memory_compare_test.c

```diff
@@ -0,0 +1,201 @@
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include <cmocka.h>
+#include <stdio.h>
+#include "ockam/memory.h"
+#include "memory/stdlib/stdlib.h"
+
+#define MAX_MEMORY_IMPLEMENTATIONS_COUNT 5
+
+typedef struct {
+  ockam_memory_t   memory[MAX_MEMORY_IMPLEMENTATIONS_COUNT];
+  size_t           implementations_count;
+} test_state_t;
+
+static int test_setup(void **state)
+{
+  static test_state_t test_state;
+
+  size_t i = 0;
+
+  // ifdef should be added here when more memory implementations added
```

Comment:
To make sure I understand it right, in this code there's an executable that is built for whatever target we compile for, but the executable exercises all memory implementations? My only concern with this is if there are certain memory implementations that can't run on a specific target we need a way for them to not be built, easy enough to do with a compiler flag or similar I suppose.

My original thought for a layout would have been:
`lib/memory/stdlib/test/test_stdlib.c` -> Contains `main()`, calls `test_compare()`
`lib/memory/test/test_compare.c` -> Contains `test_compare()`
The upside here is the stdlib directory only gets built for the targets its supported on. 

I think either solution works, we just need to make sure we have a way to disable certain implementations when compiling for certain targets. Thoughts?

Reply:
@mtm0183 I understand that at some point we'll need to enable/disable specific implementations, my plan was:
- define cmake variable, e.g. OCKAM_STD_MEMORY_ENABLED
- based on that variable we decide whether we a) are building stdlib library b) are linking it to other libraries
- tests can simply check #ifdef OCKAM_STD_MEMORY_ENABLED in setup function and add corresponding implementation to test_state object
- tests themselves don't care about all of that, they are given state object with array of implementations, and they test them, they don't even know what implementations are

Reply:
Ok. We may find as we move forward we end up with too many cmake variables for the various implementations but for now I don't think it is an issue. This is a fine solution and we can always revisit it as we add more implementations.